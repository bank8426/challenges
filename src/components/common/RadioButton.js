import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
/* The container */
const StyledLabel = styled.label`
  display: block;
  position: relative;
  padding-left: 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  
  /* On mouse-over, add a grey background color */
  &:hover input ~ .checkmark {
      background-color: #ccc;
  }

  /* When the radio button is checked, add a blue background */
  & input:checked ~ .checkmark {
      background-color: #2196F3;
  }

  /* Show the indicator (dot/circle) when checked */
  & input:checked ~ .checkmark:after {
      display: block;
  }

  /* Style the indicator (dot/circle) */
  & .checkmark:after {
    margin: 25%;
    width: 50%;
    height: 50%;
    border-radius: 50%;
    background: white;
  } 
`
/* Hide the browser's default radio button */
const StyledRadioButton = styled.input`
    position: absolute;
    opacity: 0;
    cursor: pointer;
`

/* span to display styled that represent radio button*/
const StyledCheckmark = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
  border-radius: 50%;
  -webkit-transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  /* Create the indicator (the dot/circle - hidden when not checked) */
  &:after {
    content: "";
    position: absolute;
    display: none;
  }
`

/* h3 to display text for radio button*/
const StyledH3 = styled.h3`
  display:inline;
`

/**
 * label to display radio button with styled
 */
const RadioButton = ({name,handleChange,displayMessage,checked}) => {
  return (
    <StyledLabel>
      <StyledRadioButton
        type="radio"
        name={name}
        onChange={ () => {handleChange()}} 
        checked={checked}
      />
      <StyledH3>{displayMessage}</StyledH3>
      <StyledCheckmark className="checkmark"></StyledCheckmark>
    </StyledLabel>
  )
}

RadioButton.defaultProps = {
  name:'',
  handleChange: () => {},
  displayMessage :'',
  checked:false,
}

RadioButton.propTypes={
  /** string name of radio button*/
  name:PropTypes.string.isRequired,
  /** function that will call when click on radio button*/
  handleChange: PropTypes.func.isRequired,
  /** 
  * display text that will show along radio button
  * can be number or string
  */
  displayMessage :PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  /** boolean default check state for radio button */
  checked:PropTypes.bool.isRequired,
}

export default RadioButton
