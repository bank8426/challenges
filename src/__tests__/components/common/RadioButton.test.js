import React from 'react'
import RadioButton from '../../../components/common/RadioButton'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme';
import 'jest-styled-components'

it('should display radio button', () => {
  const tree = renderer.create(<RadioButton />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should display radio with checked state', () => {
  const tree = renderer.create(<RadioButton name='radio1' handleChange={jest.fn() } displayMessage='' checked={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should handle onChange', () => {
  const mockFn = jest.fn();
  const wrapper = mount(<RadioButton name='radio1' handleChange={mockFn} displayMessage='' checked={false} />);
  wrapper.find('input').simulate('change')
  expect(mockFn).toHaveBeenCalledTimes(1)
})

it('should passed checked to the input', () => {
  const mockFn = jest.fn();
  const wrapper = mount(
    <RadioButton
      name='radio1'
      handleChange={mockFn} 
      displayMessage='100' 
      checked={true}
    />,
  ).find('input');

  expect(wrapper.prop('checked')).toEqual(true);
});

it('should passed name to the input', () => {
  const mockFn = jest.fn();
  const wrapper = mount(
    <RadioButton
      name='radio1'
      handleChange={mockFn} 
      displayMessage='100' 
      checked={false}
    />,
  ).find('input');

  expect(wrapper.prop('name')).toEqual('radio1');
});


it('should passed displayMessage to the h3', () => {
  const mockFn = jest.fn();
  const wrapper = mount(
    <RadioButton
      name='radio1'
      handleChange={mockFn} 
      displayMessage='100' 
      checked={false}
    />,
  ).find('h3');

  expect(wrapper.text()).toEqual('100');
});

it('should have default handleChange', () => {
  const wrapper = mount(
    <RadioButton
      name='radio1'
      displayMessage='100' 
      checked={false}
    />,
  );
  const props = wrapper.props()
  const spy = jest.spyOn(props, 'handleChange');
  props.handleChange()
  expect(spy).toHaveBeenCalledTimes(1)
});