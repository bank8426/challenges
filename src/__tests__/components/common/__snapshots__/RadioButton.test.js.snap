// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display radio button 1`] = `
.c0 {
  display: block;
  position: relative;
  padding-left: 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:hover input ~ .checkmark {
  background-color: #ccc;
}

.c0 input:checked ~ .checkmark {
  background-color: #2196F3;
}

.c0 input:checked ~ .checkmark:after {
  display: block;
}

.c0 .checkmark:after {
  margin: 25%;
  width: 50%;
  height: 50%;
  border-radius: 50%;
  background: white;
}

.c1 {
  position: absolute;
  opacity: 0;
  cursor: pointer;
}

.c3 {
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
  border-radius: 50%;
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
}

.c3:after {
  content: "";
  position: absolute;
  display: none;
}

.c2 {
  display: inline;
}

<label
  className="c0"
>
  <input
    checked={false}
    className="c1"
    name=""
    onChange={[Function]}
    type="radio"
  />
  <h3
    className="c2"
  >
    
  </h3>
  <span
    className="checkmark c3"
  />
</label>
`;

exports[`should display radio with checked state 1`] = `
.c0 {
  display: block;
  position: relative;
  padding-left: 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:hover input ~ .checkmark {
  background-color: #ccc;
}

.c0 input:checked ~ .checkmark {
  background-color: #2196F3;
}

.c0 input:checked ~ .checkmark:after {
  display: block;
}

.c0 .checkmark:after {
  margin: 25%;
  width: 50%;
  height: 50%;
  border-radius: 50%;
  background: white;
}

.c1 {
  position: absolute;
  opacity: 0;
  cursor: pointer;
}

.c3 {
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
  border-radius: 50%;
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
}

.c3:after {
  content: "";
  position: absolute;
  display: none;
}

.c2 {
  display: inline;
}

<label
  className="c0"
>
  <input
    checked={true}
    className="c1"
    name="radio1"
    onChange={[Function]}
    type="radio"
  />
  <h3
    className="c2"
  >
    
  </h3>
  <span
    className="checkmark c3"
  />
</label>
`;
